File: backend/src/services/cart.service.ts
import { prisma } from '../config/prisma';
import { AppError } from '../middleware/error';

export class CartService {
  private static async getOrCreateCart(cartId?: string, userId?: number) {
    if (cartId) {
      const existingCart = await prisma.cart.findUnique({
        where: { id: cartId },
        include: {
          items: {
            include: {
              product: true,
            },
            orderBy: { createdAt: 'asc' },
          },
        },
      });

      if (existingCart) {
        if (userId && existingCart.userId !== userId) {
          await prisma.cart.update({
            where: { id: existingCart.id },
            data: { userId },
          });
        }
        return existingCart;
      }
    }

    return await prisma.cart.create({
      data: {
        userId: userId || null,
      },
      include: {
        items: {
          include: {
            product: true,
          },
          orderBy: { createdAt: 'asc' },
        },
      },
    });
  }

  public static async getCart(cartId?: string, userId?: number) {
    const cart = await this.getOrCreateCart(cartId, userId);
    return this.formatCartResponse(cart);
  }

  public static async addItem(cartId: string | undefined, productId: number, quantity: number, userId?: number) {
    const product = await prisma.product.findUnique({
      where: { id: productId },
    });

    if (!product) {
      throw new AppError(`Product with ID ${productId} not found`, 404);
    }

    const cart = await this.getOrCreateCart(cartId, userId);

    const existingItem = await prisma.cartItem.findUnique({
      where: {
        cartId_productId: {
          cartId: cart.id,
          productId,
        },
      },
    });

    const newQuantity = (existingItem ? existingItem.quantity : 0) + quantity;

    if (newQuantity > product.stock) {
      throw new AppError(
        `Insufficient stock for ${product.name}. Available: ${product.stock}, Requested total: ${newQuantity}`,
        400
      );
    }

    if (existingItem) {
      await prisma.cartItem.update({
        where: { id: existingItem.id },
        data: { quantity: newQuantity },
      });
    } else {
      await prisma.cartItem.create({
        data: {
          cartId: cart.id,
          productId,
          quantity,
        },
      });
    }

    return this.getCart(cart.id);
  }

  public static async updateItemQuantity(cartId: string, productId: number, quantity: number) {
    const cart = await prisma.cart.findUnique({
      where: { id: cartId },
    });

    if (!cart) {
      throw new AppError('Cart not found', 404);
    }

    const product = await prisma.product.findUnique({
      where: { id: productId },
    });

    if (!product) {
      throw new AppError(`Product with ID ${productId} not found`, 404);
    }

    if (quantity > product.stock) {
      throw new AppError(
        `Insufficient stock for ${product.name}. Available: ${product.stock}, Requested: ${quantity}`,
        400
      );
    }

    const existingItem = await prisma.cartItem.findUnique({
      where: {
        cartId_productId: {
          cartId,
          productId,
        },
      },
    });

    if (!existingItem) {
      throw new AppError(`Product is not in your cart`, 404);
    }

    await prisma.cartItem.update({
      where: { id: existingItem.id },
      data: { quantity },
    });

    return this.getCart(cartId);
  }

  public static async removeItem(cartId: string, productId: number) {
    const existingItem = await prisma.cartItem.findUnique({
      where: {
        cartId_productId: {
          cartId,
          productId,
        },
      },
    });

    if (!existingItem) {
      throw new AppError(`Product is not in your cart`, 404);
    }

    await prisma.cartItem.delete({
      where: { id: existingItem.id },
    });

    return this.getCart(cartId);
  }

  public static async clearCart(cartId: string) {
    const cart = await prisma.cart.findUnique({
      where: { id: cartId },
    });

    if (!cart) {
      throw new AppError('Cart not found', 404);
    }

    await prisma.cartItem.deleteMany({
      where: { cartId },
    });

    return this.getCart(cartId);
  }

  private static formatCartResponse(cart: any) {
    let totalAmount = 0;
    let itemCount = 0;

    const items = cart.items.map((item: any) => {
      const priceNum = Number(item.product.price);
      const itemTotal = priceNum * item.quantity;

      totalAmount += itemTotal;
      itemCount += item.quantity;

      return {
        id: item.id,
        productId: item.productId,
        name: item.product.name,
        price: priceNum,
        imageUrl: item.product.imageUrl,
        quantity: item.quantity,
        stock: item.product.stock,
        itemTotal,
      };
    });

    return {
      cartId: cart.id,
      items,
      itemCount,
      totalAmount,
    };
  }
}

File: backend/src/main/java/com/razorpay/integration/service/CartService.java
package com.razorpay.integration.service;

import com.razorpay.integration.dto.CartItemResponse;
import com.razorpay.integration.dto.CartResponse;
import com.razorpay.integration.exception.AppException;
import com.razorpay.integration.model.Cart;
import com.razorpay.integration.model.CartItem;
import com.razorpay.integration.model.Product;
import com.razorpay.integration.model.User;
import com.razorpay.integration.repository.CartItemRepository;
import com.razorpay.integration.repository.CartRepository;
import com.razorpay.integration.repository.ProductRepository;
import com.razorpay.integration.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class CartService {

    private final CartRepository cartRepository;
    private final CartItemRepository cartItemRepository;
    private final ProductRepository productRepository;
    private final UserRepository userRepository;

    @Transactional
    public Cart getOrCreateCart(String cartId, Integer userId) {
        if (cartId != null && !cartId.trim().isEmpty()) {
            Cart existingCart = cartRepository.findById(cartId).orElse(null);
            if (existingCart != null) {
                if (userId != null && (existingCart.getUser() == null || !existingCart.getUser().getId().equals(userId))) {
                    User user = userRepository.findById(userId).orElse(null);
                    existingCart.setUser(user);
                    existingCart = cartRepository.save(existingCart);
                }
                return existingCart;
            }
        }

        User user = userId != null ? userRepository.findById(userId).orElse(null) : null;
        Cart newCart = Cart.builder()
                .id(UUID.randomUUID().toString())
                .user(user)
                .items(new ArrayList<>())
                .build();

        return cartRepository.save(newCart);
    }

    @Transactional
    public CartResponse getCart(String cartId, Integer userId) {
        Cart cart = getOrCreateCart(cartId, userId);
        return formatCartResponse(cart);
    }

    @Transactional
    public CartResponse addItem(String cartId, Integer productId, int quantity, Integer userId) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new AppException("Product with ID " + productId + " not found", HttpStatus.NOT_FOUND));

        Cart cart = getOrCreateCart(cartId, userId);

        CartItem existingItem = cartItemRepository.findByCartIdAndProductId(cart.getId(), productId).orElse(null);
        int currentQty = existingItem != null ? existingItem.getQuantity() : 0;
        int newQuantity = currentQty + quantity;

        if (newQuantity > product.getStock()) {
            throw new AppException(
                    String.format("Insufficient stock for %s. Available: %d, Requested total: %d",
                            product.getName(), product.getStock(), newQuantity),
                    HttpStatus.BAD_REQUEST
            );
        }

        if (existingItem != null) {
            existingItem.setQuantity(newQuantity);
            cartItemRepository.save(existingItem);
        } else {
            CartItem newItem = CartItem.builder()
                    .cart(cart)
                    .product(product)
                    .quantity(quantity)
                    .build();
            cartItemRepository.save(newItem);
        }

        Cart updatedCart = cartRepository.findById(cart.getId()).orElse(cart);
        return formatCartResponse(updatedCart);
    }

    @Transactional
    public CartResponse updateItemQuantity(String cartId, Integer productId, int quantity, Integer userId) {
        Cart cart = getOrCreateCart(cartId, userId);

        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new AppException("Product with ID " + productId + " not found", HttpStatus.NOT_FOUND));

        if (quantity > product.getStock()) {
            throw new AppException(
                    String.format("Insufficient stock for %s. Available: %d, Requested: %d",
                            product.getName(), product.getStock(), quantity),
                    HttpStatus.BAD_REQUEST
            );
        }

        CartItem existingItem = cartItemRepository.findByCartIdAndProductId(cart.getId(), productId)
                .orElseThrow(() -> new AppException("Product is not in your cart", HttpStatus.NOT_FOUND));

        if (quantity <= 0) {
            if (cart.getItems() != null) {
                cart.getItems().removeIf(item -> item.getId().equals(existingItem.getId()));
            }
            cartItemRepository.delete(existingItem);
        } else {
            existingItem.setQuantity(quantity);
            cartItemRepository.save(existingItem);
        }

        Cart updatedCart = cartRepository.findById(cart.getId()).orElse(cart);
        return formatCartResponse(updatedCart);
    }

    @Transactional
    public CartResponse removeItem(String cartId, Integer productId, Integer userId) {
        Cart cart = getOrCreateCart(cartId, userId);

        CartItem existingItem = cartItemRepository.findByCartIdAndProductId(cart.getId(), productId)
                .orElseThrow(() -> new AppException("Product is not in your cart", HttpStatus.NOT_FOUND));

        if (cart.getItems() != null) {
            cart.getItems().removeIf(item -> item.getId().equals(existingItem.getId()));
        }
        cartItemRepository.delete(existingItem);

        Cart updatedCart = cartRepository.findById(cart.getId()).orElse(cart);
        return formatCartResponse(updatedCart);
    }

    @Transactional
    public CartResponse clearCart(String cartId, Integer userId) {
        Cart cart = getOrCreateCart(cartId, userId);

        cartItemRepository.deleteByCartId(cart.getId());
        if (cart.getItems() != null) {
            cart.getItems().clear();
        }

        Cart updatedCart = cartRepository.findById(cart.getId()).orElse(cart);
        return formatCartResponse(updatedCart);
    }

    public CartResponse formatCartResponse(Cart cart) {
        BigDecimal totalAmount = BigDecimal.ZERO;
        int itemCount = 0;
        List<CartItemResponse> itemResponses = new ArrayList<>();

        if (cart.getItems() != null) {
            for (CartItem item : cart.getItems()) {
                BigDecimal itemPrice = item.getProduct().getPrice();
                BigDecimal itemTotal = itemPrice.multiply(BigDecimal.valueOf(item.getQuantity()));

                totalAmount = totalAmount.add(itemTotal);
                itemCount += item.getQuantity();

                itemResponses.add(CartItemResponse.builder()
                        .id(item.getId())
                        .productId(item.getProduct().getId())
                        .name(item.getProduct().getName())
                        .price(itemPrice)
                        .imageUrl(item.getProduct().getImageUrl())
                        .quantity(item.getQuantity())
                        .stock(item.getProduct().getStock())
                        .itemTotal(itemTotal)
                        .build());
            }
        }

        return CartResponse.builder()
                .cartId(cart.getId())
                .items(itemResponses)
                .itemCount(itemCount)
                .totalAmount(totalAmount)
                .build();
    }
}
